File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int begin = 0;
        int maxLen = 0;

        for (int end = 0; end < s.length(); end++) {
            while (set.contains(s.charAt(end))) {
                set.remove(s.charAt(begin));
                begin++;
            }
            int len = end - begin + 1;
            maxLen = Math.max(len, maxLen);
            set.add(s.charAt(end));
        }

        return maxLen;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.cpp
class Solution {
   public:
    int characterReplacement(string s, int k) {
        int alpha[26] = {0};
        int l = 0;
        int freq = 0;
        int max_len = 0;
        for (int r = 0; r < s.length(); r++) {
            alpha[s[r] - 'A']++;
            freq = max(freq, *max_element(alpha, alpha + 26));
            while (r - l + 1 - freq > k) {
                alpha[s[l] - 'A']--;
                l++;
            }
            max_len = max(max_len, r - l + 1);
        }
        return max_len;
    }
};
